### PR TITLE
docs/project structure technologies migrations

### PR DESCRIPTION
**Pull request by kwstnr**
- **docs(project-structure): document project structure and architecture**
- **docs(projektstruktur): add 12-factor app factors to documentation**
- **docs(projekstruktur): techniques and tools chapter**
- **docs(persitence-and-migrations): add persistence and migrations chapter**
- **docs(learnings): add my learnings**

### CODE REVIEW

File: docs/chapters/persistence-and-migrations.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Die erzeuget Migration wird im `Migration`-Ordner des Data-Projekt abgelegt. Diese umfasst aber lediglich die Schemaänderungen. Datenmigrationen müssen jeweils in der `Up` und `Down` Methode manuell implementiert werden.
+Diese müssen explizit getestet werden. Eine Infrastruktur für solche Tests existiert noch nicht. Die Verantwortung liegt dementsprechend bei jedem einzelnen Contributer und den Reviewern.
+
+### Baby Steps und Backwards Compatibility
+Migrationen sollten stets in Baby Steps ausgeführt werden. Diese sollen vor dem finalen Umstieg stets Rückwärts Kompatibel sein.
+Dementsprechend benötigt eine risikobehaftete Migration stets mehrere Versionen, bis final ein Breaking Change ausgelöst wird.
+Dies verlangt einen erhöhten Managementaufwand, ermöglicht jedoch sichere Breaking Schema Changes.
+
+### Fehlgeschlagene Mutationen
+Mutationen die Technisch nicht umsetzbar sind, sollten es nie auf die DEV Umgebung durch Integrationstests schaffen. Die Integrationstests in Question müssen aber noch umgesetzt werden...
```
Comment (Livio2000) [inline docs/chapters/persistence-and-migrations.md:20]: Question auf deutsch schreiben

File: docs/chapters/persistence-and-migrations.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+Falls eine Mutation fachlich nicht wie erwartet funktioner hat, muss diese über eine weitere Mutation in einem HotFix rückwärts gemacht werden. Auf den produktiven Deployment Stages werden keine Migrationen zurückgerollt.
+Dies bedeutet aber nicht, dass fehlgeschlagene Version-Upgrades ohne Migrationen nicht gerollbacked werden dürfen!
```
Comment (Livio2000) [inline docs/chapters/persistence-and-migrations.md:22]: zurückgerollt anstatt gerollbacked

File: docs/chapters/projektstruktur.md [hunk 3]
```diff
@@ -1,2 +1,99 @@
...
+Wie bereits angetönt, handelt es sich bei dem `devops-book-manager` um eine GraphQL API, mit welcher eine klassische Übungs-Domäne verwaltet werden kann.
+Wie es der Name des Projekts schon so sprechend definiert, handelt es sich dabei um einen Buch-Manager. Hier können Bücher, deren Authoren und Genres verwaltet werden.
+Ein paar Bücher und Authoren wurden Beispielshalber bereits in die Datenbank geseedet um den Graphen darzustellen und navigierbar zu machen.
+
+## Projektarchitektur
+Bei dem `devops-book-manager` handelt es sich um eine klassische [layered architecture](https://www.oreilly.com/library/view/software-architecture-patterns/9781491971437/ch01.html).
+Die Applikation wurde dreigeteilt in *Data Layer*, *Domain* oder *Business Layer* und das *Graph Layer*. Die einzelnen Layers werden in folge detaillierter erläutert.
+
+### Domain Layer
+Das Domain Layer beinhaltet das, in diesem Fall relativ fett- und vorallem blutarme ([anemic](https://martinfowler.com/bliki/AnemicDomainModel.html)) Domänenmodell.
```
Comment (Livio2000) [inline docs/chapters/projektstruktur.md:30]: Der Domain Layer? anstatt Das

Comment (kwstnr) [reply docs/chapters/projektstruktur.md:30]: @Livio2000 
3EAu

**Review by Livio2000 — APPROVED** [review]
LGTM

**Pull request merged**